File: anvil-spigot/src/main/java/org/anvilpowered/anvil/spigot/module/ApiSpigotModule.java
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020-2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.anvil.spigot.module;

import com.google.inject.TypeLiteral;
import net.md_5.bungee.api.chat.TextComponent;
import org.anvilpowered.anvil.api.PlatformImpl;
import org.anvilpowered.anvil.api.command.CommandExecuteService;
import org.anvilpowered.anvil.api.command.CommandService;
import org.anvilpowered.anvil.api.misc.BindingExtensions;
import org.anvilpowered.anvil.api.server.LocationService;
import org.anvilpowered.anvil.api.util.KickService;
import org.anvilpowered.anvil.api.util.PermissionService;
import org.anvilpowered.anvil.api.util.TextService;
import org.anvilpowered.anvil.api.util.UserService;
import org.anvilpowered.anvil.common.command.CommonCallbackCommand;
import org.anvilpowered.anvil.common.entity.EntityUtils;
import org.anvilpowered.anvil.common.module.JavaUtilLoggingAdapter;
import org.anvilpowered.anvil.common.module.PlatformModule;
import org.anvilpowered.anvil.spigot.command.SpigotCallbackCommand;
import org.anvilpowered.anvil.spigot.command.SpigotCommandExecuteService;
import org.anvilpowered.anvil.spigot.command.SpigotCommandService;
import org.anvilpowered.anvil.spigot.entity.SpigotEntityUtils;
import org.anvilpowered.anvil.spigot.server.SpigotLocationService;
import org.anvilpowered.anvil.spigot.util.SpigotKickService;
import org.anvilpowered.anvil.spigot.util.SpigotPermissionService;
import org.anvilpowered.anvil.spigot.util.SpigotTextService;
import org.anvilpowered.anvil.spigot.util.SpigotUserService;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

public class ApiSpigotModule extends PlatformModule {

    public ApiSpigotModule() {
        super(new PlatformImpl(false, "spigot", JavaUtilLoggingAdapter::bindLogger));
    }

    @Override
    protected void configure() {
        super.configure();
        bind(new TypeLiteral<CommonCallbackCommand<TextComponent, CommandSender>>() {
        }).toProvider(BindingExtensions.asInternalProvider(SpigotCallbackCommand.class));
        bind(CommandExecuteService.class).to(SpigotCommandExecuteService.class);
        bind(new TypeLiteral<CommandService<CommandExecutor, CommandSender>>() {
        }).to(SpigotCommandService.class);
        bind(KickService.class).to(SpigotKickService.class);
        bind(EntityUtils.class).to(SpigotEntityUtils.class);
        bind(LocationService.class).to(SpigotLocationService.class);
        bind(PermissionService.class).to(SpigotPermissionService.class);
        bind(new TypeLiteral<TextService<TextComponent, CommandSender>>() {
        }).to(SpigotTextService.class);
        bind(new TypeLiteral<UserService<Player, Player>>() {
        }).to(SpigotUserService.class);
    }
}

File: anvil-spigot/src/main/kotlin/org/anvilpowered/anvil/spigot/module/ApiSpigotModule.kt
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020-2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.anvilpowered.anvil.spigot.module

import org.anvilpowered.anvil.api.command.CommandExecuteService
import org.anvilpowered.anvil.api.command.SimpleCommandService
import org.anvilpowered.anvil.api.misc.bind
import org.anvilpowered.anvil.api.misc.to
import org.anvilpowered.anvil.api.server.LocationService
import org.anvilpowered.anvil.api.util.InfoDumpService
import org.anvilpowered.anvil.api.util.KickService
import org.anvilpowered.anvil.api.util.PermissionService
import org.anvilpowered.anvil.api.util.TextService
import org.anvilpowered.anvil.api.util.UserService
import org.anvilpowered.anvil.common.PlatformImpl
import org.anvilpowered.anvil.common.entity.EntityUtils
import org.anvilpowered.anvil.common.module.JavaUtilLoggingAdapter
import org.anvilpowered.anvil.common.module.PlatformModule
import org.anvilpowered.anvil.common.util.CommonInfoDumpService
import org.anvilpowered.anvil.common.util.CommonTextService
import org.anvilpowered.anvil.common.util.SendTextService
import org.anvilpowered.anvil.spigot.command.SpigotCommandExecuteService
import org.anvilpowered.anvil.spigot.command.SpigotSimpleCommandService
import org.anvilpowered.anvil.spigot.entity.SpigotEntityUtils
import org.anvilpowered.anvil.spigot.server.SpigotLocationService
import org.anvilpowered.anvil.spigot.util.SpigotKickService
import org.anvilpowered.anvil.spigot.util.SpigotPermissionService
import org.anvilpowered.anvil.spigot.util.SpigotSendTextService
import org.anvilpowered.anvil.spigot.util.SpigotUserService
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class ApiSpigotModule : PlatformModule(
  PlatformImpl(
    "spigot",
    false,
    { Bukkit.getVersion() },
    JavaUtilLoggingAdapter::bindLogger,
  )
) {
  override fun configure() {
    super.configure()
    with(binder()) {
      bind<CommandExecuteService>().to<SpigotCommandExecuteService>()
      bind<EntityUtils>().to<SpigotEntityUtils>()
      bind<InfoDumpService<CommandSender>>().to<CommonInfoDumpService<CommandSender>>()
      bind<InfoDumpService<*>>().to<CommonInfoDumpService<CommandSender>>()
      bind<KickService>().to<SpigotKickService>()
      bind<LocationService>().to<SpigotLocationService>()
      bind<PermissionService>().to<SpigotPermissionService>()
      bind<SendTextService<CommandSender>>().to<SpigotSendTextService>()
      bind<SendTextService<*>>().to<SpigotSendTextService>()
      bind<SimpleCommandService<CommandSender>>().to<SpigotSimpleCommandService>()
      bind<SimpleCommandService<*>>().to<SpigotSimpleCommandService>()
      bind<TextService<CommandSender>>().to<CommonTextService<CommandSender>>()
      bind<TextService<*>>().to<CommonTextService<CommandSender>>()
      bind<UserService<Player, Player>>().to<SpigotUserService>()
    }
  }
}
